toxcore/nat_traversal.c

```diff
@@ -0,0 +1,172 @@
+    switch (error) {
+        case 0:
+            LOGGER_WARNING(log, "No IGD was found.");
+            break;
+        case 1:
+            LOGGER_INFO(log, "A valid IGD has been found.");
+
+            char portstr[10];
+            snprintf(portstr, sizeof(portstr), "%d", port);
+
+            switch (proto) {
+                case NAT_TRAVERSAL_UDP:
+                    error = UPNP_AddPortMapping(urls.controlURL, data.first.servicetype, portstr, portstr, lanaddr, "Tox", "UDP", 0, "0");
+                    break;
+                case NAT_TRAVERSAL_TCP:
+                    error = UPNP_AddPortMapping(urls.controlURL, data.first.servicetype, portstr, portstr, lanaddr, "Tox", "TCP", 0, "0");
+                    break;
+                default:
+                    LOGGER_WARNING(log, "UPnP port mapping failed (unknown NAT_TRAVERSAL_PROTO)");
+                    break;
+            }
+
+            if (error)
+                LOGGER_WARNING(log, "UPnP port mapping failed (%s)", strupnperror(error));
+            else
+                LOGGER_INFO(log, "UPnP mapped port %d", port);
+            break;
+        case 2:
+            LOGGER_WARNING(log, "IGD was found but reported as not connected.");
+            break;
+        case 3:
+            LOGGER_WARNING(log, "UPnP device was found but not recoginzed as IGD.");
+            break;
+        default:
+            LOGGER_WARNING(log, "Unknown error finding IGD (%s)", strupnperror(error));
+            break;
+    }
+
+    FreeUPNPUrls(&urls);
+
+    if (local_logger)
+        logger_kill(log);
+}
+#endif
+
+
+#ifdef HAVE_LIBNATPMP
+/* Setup port forwarding using NAT-PMP */
+void natpmp_map_port(Logger *log, NAT_TRAVERSAL_PROTO proto, uint16_t port)
+{
+    int error;
+    bool local_logger = false;
+    natpmp_t natpmp;
+    natpmpresp_t resp;
+
+    if (log == NULL) {
+        log = logger_new();
+        local_logger = true;
+    }
+
+    LOGGER_DEBUG(log, "Attempting to set up NAT-PMP port forwarding");
+
+    error = initnatpmp(&natpmp, 0, 0);
+    if (error) {
+        LOGGER_WARNING(log, "NAT-PMP initialization failed (error = %d)", error);
+        return;
+    }
+
+    if (proto == NAT_TRAVERSAL_UDP)
+        error = sendnewportmappingrequest(&natpmp, NATPMP_PROTOCOL_UDP, port, port, 3600);
+    else if (proto == NAT_TRAVERSAL_TCP)
+        error = sendnewportmappingrequest(&natpmp, NATPMP_PROTOCOL_TCP, port, port, 3600);
+    else {
+        LOGGER_WARNING(log, "NAT-PMP port mapping failed (unknown NAT_TRAVERSAL_PROTO)");
+        closenatpmp(&natpmp);
+        return;
+    }
+
+    // From line 171 of natpmp.h: "12 = OK (size of the request)"
+    if (error != 12) {
+        LOGGER_WARNING(log, "NAT-PMP send request failed (error = %d)", error);
+        closenatpmp(&natpmp);
+        return;
+    }
+
+    error = readnatpmpresponseorretry(&natpmp, &resp);
+    for ( ; error == NATPMP_TRYAGAIN ; error = readnatpmpresponseorretry(&natpmp, &resp) )
```

Comment:
Again: run `other/astyle/format-source`. This code has several style violations.